### ISSUE TITLE
tuple length inference fails when size is assigned to variable

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/pyright#6216]
**Describe the bug**
Tuple length inference works in the first case, but not the second. Marking `N` `Final` does not help.

**Code or Screenshots**
```python
def passes(items: list[str]) -> tuple[str, str]:
    tuple_ = tuple(items)
    if len(tuple_) != 2:
        raise ValueError
    return tuple_


def fails(items: list[str]) -> tuple[str, str]:
    tuple_ = tuple(items)
    N = 2
    if len(tuple_) != N:
        raise ValueError
    return tuple_
```

**VS Code extension or command-line**
`pyright 1.1.332`

**Comment by erictraut** [comment]
This is intended behavior, not a bug.

The first code sample is a supported type guard pattern. The second sample is not. For a complete list of supported type guard patterns in pyright, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards). This is the type guard pattern `len(x) == L` where `x` is a tuple (or union or tuples) and `L` is a literal positive integer expression. Pyright does not support arbitrary expressions for the RHS operand.

It is theoretically possible to extend support for this type guard to allow arbitrary RHS expressions that evaluate to a literal `int` type, but this would be costly at type evaluation time. This is the reason the current implementation restricts this to an integer literal token. We could consider extending this in the future if there is sufficient evidence that it's a common use case, but for now I'm going to leave it as is.

I'll note that the current version of mypy doesn't support this type guard pattern at all, although it appears that the mypy maintainers [just added](https://github.com/python/mypy/commit/a3af87bf252f0ed0c6e0f977ad4079418b37a70f) limited support for this type guard.

**Comment by bersbersbers** [comment]
Thanks for the explanation!

> I'll note that the current version of mypy doesn't support this type guard pattern at all, although it appears that the mypy maintainers [just added](https://github.com/python/mypy/commit/a3af87bf252f0ed0c6e0f977ad4079418b37a70f) limited support for this type guard.

Yes, this is what brought me here. With the commit you mention, `mypy` does support the `N` case (at least with `Final` - I haven't checked other cases with `mypy`.)

> We could consider extending this in the future if there is sufficient evidence that it's a common use case

The reason why I hit this is that `ruff` flags uses of integer literals in comparisons, refering to PEP 8 as their motivation, see https://docs.astral.sh/ruff/rules/magic-value-comparison/ (note that their code example is not appropriate to illustrate this rule, see https://github.com/astral-sh/ruff/issues/8109 if interested). So that very use case you have in mind is one that may always be flagged by `ruff`, and the `ruff` fix will not be understanding by `pyright`.

**Comment by erictraut** [comment]
Thanks for the additional context.

I'm not convinced that ruff's default is beneficial here. I'd argue that your second code sample is less readable and harder to maintain than the first, so maybe ruff's defaults should be changed here. On the other hand, if ruff doesn't change this rule, that makes for a more compelling case for a type checker to support arbitrary expressions here.

I generally like to support all type guard patterns that mypy does. If mypy added support for the `len(x) == L` type guard pattern and supports arbitrary expressions for the RHS, that makes a more compelling case to support it in pyright as well.

**Comment by erictraut** [comment]
I've decided to implement the change, but I reserve the right to revert this change if there is evidence that it harms type evaluation performance in a measurable way in real-world code. Based on initial limited testing, it appears that it does not.

**Comment by erictraut** [comment]
This is included in pyright 1.1.333, which I just published. It will also be included in a future release of pylance.
